### PR TITLE
ISSUE-196 Collapse SAM algorithm to operate column by column rather than pixel by pixel

### PR DESCRIPTION
**Pull request by Luner**
SAM currently loops through every pixel and preforms necessary operations. This makes changes such that operations are preformed on a column by column basis.

This creates a significant increase in a performance.

### CODE REVIEW

File: pycoal/resampler.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty 
+# of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public 
+# License along with this program; if not, write to the Free 
+# Software Foundation, Inc., 51 Franklin Street, Fifth 
+# Floor, Boston, MA 02110-1301, USA.
+
+# Adapted from Spectral Library: resampling.py
```
Comment (lewismc) [inline pycoal/resampler.py]: Please provide actual hyperlink to the original source.

Comment (Luner) [reply pycoal/resampler.py]: Just pushed the change

**Comment by lewismc** [comment]
Thanks, this seems to have introduced a host of issues in DeepSource. If you scope and resolve that would be great. Thank you.

**Comment by Luner** [comment]
I took a look earlier and the deepsource issues are largely maintained within the resampler.py adapted from the spectral code.  I can look into modifying the file such that they pass through the deepsource issues.

**Comment by Luner** [comment]
The nosetests seem to fail due to the addition of pytorch:
"ImportError: No module named torch"

I changed resampler.py such that there should be less deepsource issues.  I was wondering if I should continue modifying the file such that there are no longer deepSource issues?

**Comment by lewismc** [comment]
> I changed resampler.py such that there should be less deepsource issues. 

Excellent

> I was wondering if I should continue modifying the file such that there are no longer deepSource issues?

Yes please

**Comment by lewismc** [comment]
@Luner I fixed all DeepSource issues bar one. If you could please clean up it would be appreciated. This is then nearly ready to ship.

**Comment by Luner** [comment]
Sounds good, I should be able to get to it later today.

**Comment by lewismc** [comment]
Hi @Luner based on most recent updates to this branch, when I try to run over several new scenes, I now get the following
```
python3 example_mineral.py
Unable to parse bad band list (bbl) in header as integers.
Saving /Users/lmcgibbn/Downloads/pycoal/examples/ang20170811t220814_corr_v2p9_img_rgb.img
Unable to parse bad band list (bbl) in header as integers.
Traceback (most recent call last):
  File "example_mineral.py", line 175, in <module>
    sys.exit(main())
  File "example_mineral.py", line 147, in main
    raise e
  File "example_mineral.py", line 140, in main
    run_mineral(image, slib)
  File "example_mineral.py", line 83, in run_mineral
    mineral_classification.classify_image(input_filename, classified_filename)
  File "/usr/local/lib/python3.7/site-packages/pycoal-0.5.2-py3.7.egg/pycoal/mineral.py", line 341, in classify_image
    self.algorithm(image_file_name, classified_file_name, **self.args)
  File "/usr/local/lib/python3.7/site-packages/pycoal-0.5.2-py3.7.egg/pycoal/mineral.py", line 110, in SAM
    [x / 1000 for x in image.bands.centers], library.bands.centers)
  File "/usr/local/lib/python3.7/site-packages/pycoal-0.5.2-py3.7.egg/pycoal/resampler.py", line 103, in create_resampling_matrix
    shape[i, matches[k]] = contribs[k]
IndexError: too many indices for array
```
Can you reproduce?
Try scene for **_ang20170811t215536_**

**Comment by lewismc** [comment]
Also, can you confirm that RGB product generation seems to be busted? Only white and black pixels present in generated raster.

**Comment by Luner** [comment]
I was able to reproduce and pushed something that should fix the issue

**Comment by Luner** [comment]
I had the same problem with the RGB with both the orthonormal and reflective data.  I am going to check if the pixels are being seen as "dead" or if the bands do not exist / if there are alternative bands for the rgb values that slightly differ.

**Comment by Luner** [comment]
The hdr files mention different wavelength measurements.  I can try a few different values and see if that is part of the problem.

**Comment by Luner** [comment]
Sorry, I realize I didnt push the fix i mentioned earlier, doing that now.

**Comment by lewismc** [comment]
@Luner please wait until @mjn0898 has rebased and we have merged his branch and then rebase against master. 

@mjn0898 this creates a direct dependency on your work... meaning that us merging #199 creates a blocker for further work moving forward.

**Comment by mjn0898** [comment]
@lewismc I have resolved the conflict with the master branch.

**Comment by lewismc** [comment]
Thanks @mjn0898 

@Luner basically, please merge with master and make sure that your branch can be invoked by the new configuration parameters. Thanks

**Comment by Luner** [comment]
Got it, ill do my best to get to it later today or early tomorrow

**Comment by Luner** [comment]
So there is an issue where when running the example from within the examples directory, the file "config.ini" cannot be found.

I am changing it to "../pycoal/config.ini" However if you run mineral.py from a directory that does not conform to this path, you will have the same error:

`FileNotFoundError: [Errno 2] No such file or directory: 'config.ini'`

**Comment by Luner** [comment]
In addition to the comment above, a few things to point out about this merge:

1. There was a reintroduction of deep source issues, mostly involving the introduction of the SAM_serial function

2. There are now 3 SAM algorithms and as of now they each have the same comment headers.  It may be nice to abstract these out and only mention the differences between them.

3. It would make sense to change the joblib loop to match that of ISSUE-193 as it still uses numpy and runs in just over an hour (significant performance increase over what the joblib implementation was originally based on)

Note that the SAM_serial is currently running the code from the branch: ISSUE-193.  This is the code that still uses numpy and processes pixel by pixel.

Other than those comments, everything should be functional and ready to be merged.

**Comment by lewismc** [comment]
> So there is an issue where when running the example from within the examples directory, the file "config.ini" cannot be found.

Yes this needs to be changed to `os.path.join(os.path.dirname(__file__), file)` @mjn0898 please send in a PR for this and also provide a test to make sure it is working as expected.

> There was a reintroduction of deep source issues, mostly involving the introduction of the SAM_serial function

These need to be resolved. If this was caused by me then I apologize.

> There are now 3 SAM algorithms and as of now they each have the same comment headers. It may be nice to abstract these out and only mention the differences between them.

I complete agree. This is part of the documentation I need everyone to put together. @aheermann please take note of this and begin working on a PR to address your portion of this work. 

> It would make sense to change the joblib loop to match that of ISSUE-193 as it still uses numpy and runs in just over an hour

@aheermann please take a look at this. 

> Note that the SAM_serial is currently running the code from the branch: ISSUE-193. This is the code that still uses numpy and processes pixel by pixel.

We want to find optimizations everywhere. Please take a look at this when you can however I would like you @Luner to focus on making the pytorch implementation as optimized as possible.

> Other than those comments, everything should be functional and ready to be merged.

Lets get the DeepSource and TravisCI builds passing successfully first then we can think of merging. I apologize that this branch was merged last and it introduced more work... that just happens sometimes when work is going in parallel. I take responsibility for that. It's something which is impossible to avoid, especially in small codebases like this one.

**Comment by lewismc** [comment]
@Luner the config issues have been fixed now, please go ahead and rebase against master branch. Thank you

**Comment by lewismc** [comment]
@Luner I think most of your code in this branch has been merged into master via #203 
Please confirm and then rebase anything additional to master branch. Again, I apologize for this. It's  been a bit messy as the scope of other tickets has expended.
I suppose it's a testament to your good approach to optimizations :)

**Comment by Luner** [comment]
@lewismc It looks like the changes to the serial approach are now in master but it is still missing the pytorch implementation.  I can start a rebase and make sure everything is running smoothly

**Comment by lewismc** [comment]
Thanks
-- 

*Lewis*
Dr. Lewis J. McGibbney Ph.D, B.Sc
*Skype*: lewis.john.mcgibbney

**Comment by lewismc** [comment]
@Luner can you please take a look at resolving the conflicts in this file before we close out the project? I hope it is no longer than an hour or two work. Thanks in advance.

**Comment by Luner** [comment]
@lewismc Of course, I should be able to finish cleaning up the conflicts at tomorrow afternoon.

**Comment by lewismc** [comment]
Excellent @Luner

**Comment by Luner** [comment]
I merged the branch with master.  It looks like pytorch, serial, and joblib are all running for SAM.

**Comment by Luner** [comment]
It looks like travis-ci is having trouble locating some of the directories: "config.ini"

**Comment by Luner** [comment]
@lewismc After completing the merge, mineral.py started to seem very cluttered and confusing.  I opened a new issue #212 and pushed some changed in there as I was not sure if it was something you were looking for.

I think the changes make it much easier to see the difference between the implementation as well as remove any unnecessary clutter.

I will open a pull request for that branch with a little more detail.

**Comment by lewismc** [comment]
Thanks @Luner for updating this looks great. If you are able to address the DeepSource issues that would be a bonus otherwise I can do that please let me know. 

> It looks like travis-ci is having trouble locating some of the directories: "config.ini"

The first thing is to add `tqdm` to the list of dependencies in `setup.py`

**Pull request closed without merging**